libraries/exoplayer/src/main/java/androidx/media3/exoplayer/audio/AudioCapabilities.java

```diff
@@ -102,25 +100,33 @@ public static AudioCapabilities getCapabilities(Context context) {
     if (Util.SDK_INT >= 23 && Api23.isBluetoothConnected(context)) {
       return DEFAULT_AUDIO_CAPABILITIES;
     }
+    ImmutableSet.Builder supportedEncodings = new ImmutableSet.Builder<>();
+
     if (deviceMaySetExternalSurroundSoundGlobalSetting()
         && Global.getInt(context.getContentResolver(), EXTERNAL_SURROUND_SOUND_KEY, 0) == 1) {
-      return EXTERNAL_SURROUND_SOUND_CAPABILITIES;
+      supportedEncodings.addAll(Ints.asList(EXTERNAL_SURROUND_SOUND_ENCODINGS));
     }
     // AudioTrack.isDirectPlaybackSupported returns true for encodings that are supported for audio
     // offload, as well as for encodings we want to list for passthrough mode. Therefore we only use
     // it on TV and automotive devices, which generally shouldn't support audio offload for surround
     // encodings.
     if (Util.SDK_INT >= 29 && (Util.isTv(context) || Util.isAutomotive(context))) {
-      return new AudioCapabilities(
-          Api29.getDirectPlaybackSupportedEncodings(), DEFAULT_MAX_CHANNEL_COUNT);
+      supportedEncodings.addAll(Ints.asList(Api29.getDirectPlaybackSupportedEncodings()));
     }
+
     if (intent == null || intent.getIntExtra(AudioManager.EXTRA_AUDIO_PLUG_STATE, 0) == 0) {
-      return DEFAULT_AUDIO_CAPABILITIES;
+      if (supportedEncodings.build().isEmpty()) {
```

Comment:
1. How do you think if we handle the case when HDMI is connected first, since currently the non-HDMI case has more nested logics. For example:

```
if (intent != null) && intent.getIntExtra(AudioManager.EXTRA_AUDIO_PLUG_STATE, 0) == 1 {
    // HDMI case. Add the encodings from AudioManager.EXTRA_ENCODINGS, and return the audio capabilities based on the supportedEncodings and AudioManager.EXTRA_MAX_CHANNEL_COUNT
}

// Non HDMI case

```
2. For non-HDMI case, could we build the supportedEncodings set **just for once**, and check if it is empty, and then convert the **built set** to array to construct audio capabilities?

Reply:
@tianyif I think it is better to check getDirectPlaybackSupportedEncodings() before checking connected HDMI. I remember coming across cases where the device do not report ENCODING_DTS when it is connected to an AVR that supports it. Yet it is able to playback because of Direct passthrough playback. The SOC takes care of everything including transcoding and sending out the correct HDMI bitstream to the AVR. If we return immediately after checking the connected HDMI (as the first check), it will not know that ENCODING_DTS is actually supported by direct passthrough playback. It may revert to using mediacodec DTS decoder and then fail to transcode. The output could be PCM even though it is connected to an AVR. What do you think?

Reply:
@tianyif The way I am looking at this is:  getCapabilities() should return the union of all the encodings that are supported by direct playback and connected audio sink devices. Let me know if you think otherwise.

Reply:
Oh sorry for the confusion! I mean after all the other checks remain in their current place, and then just make the logics in the line 126 - 129 (HDMI case) **inside** of the if condition, and make the logics in the lines 118 - 123 **after** the if condition, that is:

```
// Bluetooth case, External surround sound settings case, direct playback case remain in their current place

if (intent != null) && intent.getIntExtra(AudioManager.EXTRA_AUDIO_PLUG_STATE, 0) == 1 {
    // HDMI case. Add the encodings from AudioManager.EXTRA_ENCODINGS, and return the audio capabilities based on the supportedEncodings and AudioManager.EXTRA_MAX_CHANNEL_COUNT
}

// No HDMI case, the final return
```

Reply:
@tianyif No worries. Think I mis-interpreted your intentions as well. I agree, better to swap those 2 blocks of code. Please see df0058109857f35c36e4366c549ff1f26c43d04a  Thanks.